File: impl_java/modelGeneration/Salient.java
package modelGeneration;


public record Salient<T>(T obj, Double salience) implements Comparable<Salient<T>> {

    public static final double FULL = 1.0;
    public static final double DECAY_RATE = 0.9;

    @Override
    public int compareTo(Salient<T> tSalient) {
        return salience.compareTo(tSalient.salience);
    }

    @Override
    public boolean equals(Object other) {
        return other instanceof Salient s && s.obj.equals(obj);
    }


    public Salient<T> copy() {
        return new Salient<>(obj, salience);
    }

    @Override
    public int hashCode() {
        return obj.hashCode();
    }

}

File: python_port/src/nls_python/model_generation/salient.py
from __future__ import annotations

from dataclasses import dataclass
from functools import total_ordering
from typing import ClassVar, Generic, TypeVar

from nls_python.typing_utils import typechecked

T = TypeVar("T")


@typechecked
@total_ordering
@dataclass(frozen=True)
class Salient(Generic[T]):
    """Mirror of the Java Salient record with ordering by salience."""

    obj: T
    salience: float

    FULL: ClassVar[float] = 1.0
    DECAY_RATE: ClassVar[float] = 0.9

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Salient):
            return NotImplemented
        return self.obj == other.obj

    def __lt__(self, other: "Salient[T]") -> bool:
        if not isinstance(other, Salient):
            return NotImplemented
        return self.salience < other.salience

    def __hash__(self) -> int:
        return hash(self.obj)

    def copy(self) -> "Salient[T]":
        return Salient(self.obj, self.salience)
